utils/compilation_filter.py

```diff
@@ -0,0 +1,265 @@
+
+
+class DottyPrinter:
+    """A class for generating the dotty graph file"""
+
+    def __init__(self):
+        self.vertices_ = []
+        self.edges_ = []
+
+    def get_color(self, isDirectTrans: bool) -> str:
+        """Returns the color for the given node. """
+
+        if isDirectTrans:
+            return "Yellow2"
+        else:
+            return "AliceBlue"
+
+    def dump_label(self, tran: Transformation) -> str:
+        """Returns the string for the label of the given transformation. """
+
+        labelStr = rf"""{{ {{SCOPE:\l{tran.scopeName_} }}|{{ORIGINAL NODES:\l\l"""
+        for rstr in tran.removedNodes_:
+            labelStr += rf"""{rstr}\l\l"""
+        labelStr += rf"}}| {{REPLACED BY:\l\l"
+        for astr in tran.addedNodes_:
+            labelStr += rf"""{astr}\l\l"""
+        labelStr += f"}} }}"
+        return labelStr
+
+    def dump_node(self, tran: Transformation) -> None:
+        """Generates the dotty information for the given transformation. """
+
+        if not tran:
+            return
+
+        tranStr = f"""v{tran.transID_}[\n
+                    \tlabel = \"{self.dump_label(tran)}\"\n
+                    \tshape = \"record\"\n
+                    \tstyle=\"filled,rounded\"\n
+                    \tfillcolor={self.get_color(tran.isDirectTrans_)}\n
+                    penwidth = 2];\n"""
+        self.vertices_.append(tranStr)
+
+    def visit_nodes(self) -> None:
+        """Visits all transformation and dump the dotty information for each transformation. """
+
+        for tran in TRANS_LIST:
+            self.dump_node(tran)
+
+    def visit_edges(self) -> None:
+        """Visits all edges and dump the dotty information for each edge. """
+
+        for tran in TRANS_LIST:
+            for anc in tran.ancestors_:
+                edgeStr = f"v{anc.transID_} -> v{tran.transID_}"
+                self.edges_.append(edgeStr)
+
+    def dump_graph(self) -> None:
+        """Visits the graph and generates the dotty information. """
+
+        self.visit_nodes()
+        self.visit_edges()
+        with open(f"transformations.dot", "w") as f:
+            f.write("digraph DAG {\n\trankdir=TB;\n")
+            for v in self.vertices_:
+                f.write(f"{v}\n")
+            for e in self.edges_:
+                f.write(f"{e};\n")
+            f.write("}")
+
+
+def dump_dotty_DAG():
+    """A helper function to dump the dotty file."""
+
+    dotty = DottyPrinter()
+    dotty.dump_graph()
+
+
+def parse_args():
+    """Parse the arguments of this script. """
+
+    parser = argparse.ArgumentParser(
+        description="Filter compilation and optimiztion.")
+    parser.add_argument("--db-file")
+    parser.add_argument("--filter-target")
+    options = parser.parse_args()
+
+    assert options.db_file and options.filter_target, "Please specify db file and filter target."
+    return options.db_file, options.filter_target
+
+
+def init_db(sqliteFile: str) -> sqlite3.Connection:
```

Comment:
Consider putting DB-related functions into a separate module so that they can be shared by all tools that use them.

Reply:
sounds reasonable. I'm prone to submit a separate following pr to make the changes that put different functions into different modules.